### PR TITLE
Initial release of the cash shop sales

### LINKED ISSUE #1229: Missing Cash Shop Feature: Special Tab "Update Quantity"
The current official (new) Cash Shop interface and it's used item_cash_db.txt gives us **0** possibilities to make anything dynamic at all.

Maybe i'm unaware about some specific features, but as of now i wasn't able to find anything even related to the cash shop system that could be usefull.

So the only settings we've got are:
_// Type,ItemID,Price_

**But there's no way we could possibly set a time limited offer nor anything "Special" about the special section itself such as limited quantities or whatever it has been created for.**

It would be more than great to be able to dynamicaly change/update the cash shop without the need to reload the item database.
So we can "per script" set special prices, limited offers and update the shop on the fly without accessing the server itself.

At the current state, using the old cash shop system gives us way more possibilities on updating the shops itself.
There, it's limited to our scripting skills.

But the "new" Cash Shop Button / Interface Shop **is lacking it's very core features.**

### PR DESCRIPTION
**Pull request by Lemongrass3110**
Fixes #1229

Merry christmas guys! :-)

### CODE REVIEW

**Comment by admkakaroto** [comment]
How to use this? '-'

**Comment by Atemo** [comment]
Forum discussion https://rathena.org/board/topic/108792-cash-shop-sale-system/

**Comment by admkakaroto** [comment]
@Atemo tnx!

File: conf/inter_athena.conf [hunk 1]
```diff
@@ -143,6 +143,7 @@ mob_skill_db_re_db: mob_skill_db_re
 mob_skill_db2_db: mob_skill_db2
 //mob_skill_db2_db: mob_skill_db2_re
 mapreg_table: mapreg
+sales_tabe: sales
```
Comment (secretdataz) [inline conf/inter_athena.conf]: tabe -> table

**Comment by anacondaqq** [comment]
best of the best =) Thank you a lot for release!

**Comment by Angelic234** [comment]
Step 1: Set up an item to item_cash_db.txt (ex 8,512,99;)
2: start server
3: use SQL add step.1 item to sell
4. reloaditemdb
5. buy all Quantity of Step 3 item you set up
6. waitting for Endtime
7. Map Crash

**Comment by Angelic234** [comment]
Step 1: Set up an item to item_cash_db.txt (ex 8,512,99;)
2: start server
3: use SQL add step.1 item to sell
4. reloaditemdb
5. waitting for item sell
6. then reloaditemdb again
7. Map-server => [Error]: Memory manager: freed-data is changed.
8. reloaditemdb again
9. Map-server crash

**Comment by Angelic234** [comment]
A question
We have a SQL table [price] , but never use it ?

**Comment by Lemongrass3110** [comment]
@Angelic234 removed the column just now. This one was a leftover from development. :-)

**Comment by Angelic234** [comment]
1. Set up multiples items to sell (ex. five items)
2. Buy all Quantity of first item
3. Logout and login
4. Player only can see item of 3 4 5.
5. reloaditemdb
6. Player can see item of 2 3 4 5.

File: src/map/cashshop.c [hunk 2]
```diff
@@ -139,16 +143,314 @@ static int cashshop_read_db_sql( void ){
...
+
+	return true;
+}
+
+static void sale_read_db_sql( void ){
+	uint32 lines = 0, count = 0;
+
+	if( SQL_ERROR == Sql_Query( mmysql_handle, "SELECT `nameid`, UNIX_TIMESTAMP(`start`), UNIX_TIMESTAMP(`end`), `amount` FROM `%s` WHERE `end` > now()", sales_table ) ){
+		Sql_ShowDebug(mmysql_handle);
+		return;
```
Comment (lighta) [inline src/map/cashshop.c]: I don't remenber but you don't need FreeResult ?? here

File: src/map/cashshop.c [hunk 3]
```diff
@@ -139,16 +143,314 @@ static int cashshop_read_db_sql( void ){
...
+	ShowStatus( "Done reading '"CL_WHITE"%lu"CL_RESET"' entries in '"CL_WHITE"%s"CL_RESET"'.\n", count, sales_table );
+}
+
+static int sale_end_timer( int tid, unsigned int tick, int id, intptr_t data ){
+	struct sale_item_data* sale_item = (struct sale_item_data*)data;
+
+	clif_sale_end( sale_item, NULL, ALL_CLIENT );
+
+	// Remove the timer so the sale end is not sent out again
+	delete_timer( sale_item->timer_end, sale_end_timer );
```
Comment (lighta) [inline src/map/cashshop.c]: I'd delete the timer, before sending to all_client. It just look more secure. If you have tons of client or the broadcast is too slow you could have this rescheduled before the delete. (a strech I know but still)

File: src/map/cashshop.c [hunk 4]
```diff
@@ -236,6 +567,7 @@ bool cashshop_buylist( struct map_session_data* sd, uint32 kafrapoints, int n, u
 	for( i = 0; i < n; ++i ){
 		unsigned short nameid = *( item_list + i * 5 );
 		uint32 quantity = *( item_list + i * 5 + 2 );
+		uint16 tab = *(item_list + i * 5 + 4);
```
Comment (lighta) [inline src/map/cashshop.c]: won't this be out of range for older client ?

Comment (Lemongrass3110) [inline src/map/cashshop.c]: No it was always there to begin with(`i * 5`). Only mistake here is that I used uint16 instead of uint8

**Pull request merged**